File: web/src/lib/system-widgets.ts
/**
 * System widgets — pre-defined Query notes that ship with v9. Created on
 * first app load if missing. The Phase 9.0 rail nav (Today / Pages / Timeline /
 * Graph / Properties) is reframed here as Query widgets so the rail is purely
 * data-driven from 9.1 onward.
 *
 * Widgets that don't have real middle-column behavior yet (Calendar, Inbox)
 * still ship as Query notes — their middle-column renderers stub to
 * "Coming in 9.x" until the corresponding phase lands.
 */
import type { api as Api } from "$lib/api-client";

type SystemWidgetSpec = {
  id: string;
  title: string;
  query: string;
  group?: string;
  sort?: string;
  icon?: string;
  color?: string;
  section: "pinned" | "browse" | "saved";
};

export const SYSTEM_WIDGETS: SystemWidgetSpec[] = [
  // Pinned — daily-driver entry points.
  {
    id: "today",
    title: "Today",
    query: "",
    section: "pinned",
    icon: "calendar",
    color: "amber",
  },
  {
    id: "pages",
    title: "Pages",
    query: "kind:page",
    section: "pinned",
    icon: "cal",
    color: "amber-2",
  },
  // Browse — corpus-wide views.
  {
    id: "tasks",
    title: "Tasks",
    query: "kind:block tag:Task -status:done",
    group: "status",
    sort: "deadline asc",
    section: "browse",
    icon: "task",
    color: "rose",
  },
  {
    id: "projects",
    title: "Projects",
    query: "kind:page note_type:Project",
    section: "browse",
    icon: "project",
    color: "indigo",
  },
  {
    id: "people",
    title: "People",
    query: "kind:page note_type:Person",
    section: "browse",
    icon: "person",
    color: "plum",
  },
  {
    id: "inbox",
    title: "Inbox",
    // 9.2 will refine to "no parent project, not daily, no status".
    query: "kind:block -has:status",
    section: "browse",
    icon: "inbox",
    color: "teal",
  },
  {
    id: "calendar",
    title: "Calendar",
    // 9.2 lights this up with the mini-calendar UI; for now it's a
    // listing of dated blocks.
    query: "kind:block has:scheduled",
    sort: "scheduled asc",
    section: "browse",
    icon: "cal",
    color: "amber-2",
  },
  // Saved — quick filters
  {
    id: "recent",
    title: "Recent",
    query: "kind:page",
    sort: "modified desc",
    section: "saved",
    icon: "clock",
    color: "ochre",
  },
  {
    id: "pinned",
    title: "Pinned",
    query: "kind:page",
    section: "saved",
    icon: "pin",
    color: "rose",
  },
];

function widgetTemplate(w: SystemWidgetSpec): string {
  const lines = [
    "---",
    `title: "${w.title}"`,
    `type: "Query"`,
    "tags: []",
    "---",
    `query:: ${w.query}`,
  ];
  if (w.group) lines.push(`group:: ${w.group}`);
  if (w.sort) lines.push(`sort:: ${w.sort}`);
  if (w.icon) lines.push(`icon:: ${w.icon}`);
  if (w.color) lines.push(`color:: ${w.color}`);
  lines.push(`section:: ${w.section}`);
  lines.push(""); // trailing newline
  return lines.join("\n");
}

/**
 * Idempotent: ensure every system widget exists as a Query note. Safe to call
 * on every app load. Per-widget `getNote` lookup; only creates on 404.
 *
 * `apiClient` is injected so callers can pass the singleton from `api-client.ts`
 * without coupling this module to it.
 */
export async function ensureSystemWidgets(
  apiClient: typeof Api,
): Promise<{ created: string[]; existing: string[] }> {
  const created: string[] = [];
  const existing: string[] = [];
  await Promise.all(
    SYSTEM_WIDGETS.map(async (w) => {
      try {
        await apiClient.getNote(w.id);
        existing.push(w.id);
      } catch (err) {
        // 404 (or similar) → create. Anything else, swallow — we don't want
        // a transient backend hiccup to block app load.
        try {
          await apiClient.createNote(w.title, widgetTemplate(w));
          created.push(w.id);
        } catch {
          // give up silently; user can re-create manually via ⌘K.
        }
      }
    }),
  );
  return { created, existing };
}

File: crates/tesela-core/src/system_widgets.rs
//! Default Query widgets that ship with every new mosaic — keeps the
//! rail nav (Dailies / Pages / Tasks / Projects / People / Inbox /
//! Calendar / Recent / Pinned) populated from the moment a fresh
//! mosaic is created.
//!
//! Kept in sync with `web/src/lib/system-widgets.ts` — that file
//! continues to auto-create missing widgets on web mount as a safety
//! net (e.g. a user manually deleted one). Server-side seeding here
//! eliminates the race where the rail rendered before the web's
//! idempotent `ensureSystemWidgets` had a chance to run.

use std::fs;
use std::io;
use std::path::Path;

pub struct SystemWidget {
    pub id: &'static str,
    pub title: &'static str,
    pub query: &'static str,
    pub group: Option<&'static str>,
    pub sort: Option<&'static str>,
    pub icon: &'static str,
    pub color: &'static str,
    pub section: &'static str,
}

/// Mirror of `SYSTEM_WIDGETS` in the web client. Source of truth for
/// both initial seeding (server side, runs on mosaic create) and the
/// idempotent ensure pass (web client, runs on every mount).
pub const SYSTEM_WIDGETS: &[SystemWidget] = &[
    SystemWidget {
        id: "dailies",
        title: "Dailies",
        query: "",
        group: None,
        sort: None,
        icon: "calendar",
        color: "amber",
        section: "pinned",
    },
    SystemWidget {
        id: "pages",
        title: "Pages",
        query: "kind:page",
        group: None,
        sort: None,
        icon: "cal",
        color: "amber-2",
        section: "pinned",
    },
    SystemWidget {
        id: "tasks",
        title: "Tasks",
        query: "kind:block tag:Task -status:done",
        group: Some("status"),
        sort: Some("deadline asc"),
        icon: "task",
        color: "rose",
        section: "browse",
    },
    SystemWidget {
        id: "projects",
        title: "Projects",
        query: "kind:page note_type:Project",
        group: None,
        sort: None,
        icon: "project",
        color: "indigo",
        section: "browse",
    },
    SystemWidget {
        id: "people",
        title: "People",
        query: "kind:page note_type:Person",
        group: None,
        sort: None,
        icon: "person",
        color: "plum",
        section: "browse",
    },
    SystemWidget {
        id: "inbox",
        title: "Inbox",
        query: "kind:block -has:status",
        group: None,
        sort: None,
        icon: "inbox",
        color: "teal",
        section: "browse",
    },
    SystemWidget {
        id: "calendar",
        title: "Calendar",
        query: "kind:block has:scheduled",
        group: None,
        sort: Some("scheduled asc"),
        icon: "cal",
        color: "amber-2",
        section: "browse",
    },
    SystemWidget {
        id: "recent",
        title: "Recent",
        query: "kind:page",
        group: None,
        sort: Some("modified desc"),
        icon: "clock",
        color: "ochre",
        section: "saved",
    },
    SystemWidget {
        id: "pinned",
        title: "Pinned",
        query: "kind:page",
        group: None,
        sort: None,
        icon: "pin",
        color: "rose",
        section: "saved",
    },
];

fn render(w: &SystemWidget) -> String {
    let mut out = String::new();
    out.push_str("---\n");
    out.push_str(&format!("title: \"{}\"\n", w.title));
    out.push_str("type: \"Query\"\n");
    out.push_str("tags: []\n");
    out.push_str("---\n");
    out.push_str(&format!("query:: {}\n", w.query));
    if let Some(g) = w.group {
        out.push_str(&format!("group:: {}\n", g));
    }
    if let Some(s) = w.sort {
        out.push_str(&format!("sort:: {}\n", s));
    }
    out.push_str(&format!("icon:: {}\n", w.icon));
    out.push_str(&format!("color:: {}\n", w.color));
    out.push_str(&format!("section:: {}\n", w.section));
    out
}

/// Seed `notes/<id>.md` for every system widget that doesn't already
/// exist. Idempotent — re-running is safe; a user-edited widget is
/// preserved.
pub fn seed(mosaic_root: &Path) -> io::Result<usize> {
    let notes_dir = mosaic_root.join("notes");
    fs::create_dir_all(&notes_dir)?;
    let mut created = 0;
    for w in SYSTEM_WIDGETS {
        let path = notes_dir.join(format!("{}.md", w.id));
        if path.exists() {
            continue;
        }
        fs::write(&path, render(w))?;
        created += 1;
    }
    Ok(created)
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[test]
    fn seed_creates_all_widgets_first_time() {
        let temp = TempDir::new().unwrap();
        let n = seed(temp.path()).unwrap();
        assert_eq!(n, SYSTEM_WIDGETS.len());
        // Spot-check a couple of widgets landed.
        let dailies = fs::read_to_string(temp.path().join("notes/dailies.md")).unwrap();
        assert!(dailies.contains("title: \"Dailies\""));
        assert!(dailies.contains("section:: pinned"));
        let tasks = fs::read_to_string(temp.path().join("notes/tasks.md")).unwrap();
        assert!(tasks.contains("group:: status"));
    }

    #[test]
    fn seed_is_idempotent_and_preserves_user_edits() {
        let temp = TempDir::new().unwrap();
        seed(temp.path()).unwrap();
        // User edits the dailies widget.
        let dailies_path = temp.path().join("notes/dailies.md");
        fs::write(&dailies_path, "USER EDITED\n").unwrap();
        let n = seed(temp.path()).unwrap();
        assert_eq!(n, 0, "second seed should be a no-op");
        assert_eq!(fs::read_to_string(dailies_path).unwrap(), "USER EDITED\n");
    }
}
